### PR TITLE
new loadout functions

### PR DESCRIPTION
**Pull request by TimGoll**
As an alternative to #374 I propose this hookless loadout function thingy. Actually it is way superior to hooks, since it makes it even more easy to create roles!

```lua
function ROLE:GiveRoleLoadout(ply, isRoleChange)
	InitRoleMarker(ply)
end

function ROLE:RemoveRoleLoadout(ply, isRoleChange)
	DeinitRoleMarker(ply)
end
```
That's it. No more role checks, roundchecks, etc.

### CODE REVIEW

**Comment by TimGoll** [comment]
I looked through the code and couldn't find anything else related to default loadout. `ROLE.defaultEquipment` is not the default loadout, but the default shopitems as far as I know. Additionally there is the way the supervillain used to do it and I think nobody should ever think of using this way.

```lua
hook.Add("TTT2RolesLoaded", "AddCrystalKnifeToDefaultLoadout", function()
	local wep = weapons.GetStored("weapon_ttt_crystalknife")
	if wep then
		wep.InLoadoutFor = wep.InLoadoutFor or {}

		if not table.HasValue(wep.InLoadoutFor, ROLE_SUPERVILLAIN) then
			table.insert(wep.InLoadoutFor, ROLE_SUPERVILLAIN)
		end
	end
end)
```

File: gamemodes/terrortown/gamemode/server/sv_player.lua [hunk 1]
```diff
@@ -896,6 +903,13 @@ function GM:PlayerDeath(victim, infl, attacker)
 
 	timer.Simple(0, function()
 		if IsValid(victim) then
+			-- a hook to handle the rolespecific stuff that should be done on
+			-- rolechange and respawn (while a round is active)
+			if victim:IsActive() then
+				local role = roles.GetByIndex(victim:GetSubRole())
+				role:RemoveRoleLoadout(victim, false)
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_player.lua]: victim:GetSubRoleData():RemoveRoleLoadout(victim, false) could simplify it

File: gamemodes/terrortown/gamemode/server/sv_player.lua [hunk 2]
```diff
@@ -120,6 +120,13 @@ function GM:PlayerSpawn(ply)
 	ply:SetupHands()
 
 	SCORE:HandleSpawn(ply)
+
+	-- a hook to handle the rolespecific stuff that should be done on
+	-- rolechange and respawn (while a round is active)
+	if ply:IsActive() then
+		local role = roles.GetByIndex(ply:GetSubRole())
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/server/sv_player.lua]: ply:GetSubRoleData():GiveRoleLoadout(ply, false) could simplify it

File: lua/terrortown/entities/roles/ttt_role_base/shared.lua [hunk 3]
```diff
@@ -93,3 +93,25 @@ function ROLE:GetSubRoles()
...
+	---
+	-- Function that is overwritten by the role and is called on rolechange and death.
+	-- It is used to remove the rolespecific loadout.
+	-- @param PLAYER ply
+	-- @param boolean isRoleChange This is true for a rolechange, but not for death
+	-- @realm server
+	function ROLE:RemoveRoleLoadout(ply, isRoleChange)
+
+	end
+end
```
Comment (Alf21) [inline lua/terrortown/entities/roles/ttt_role_base/shared.lua]: Missing EOL

File: gamemodes/terrortown/gamemode/shared/sh_player_ext.lua [hunk 4]
```diff
@@ -137,6 +137,13 @@ function plymeta:SetRole(subrole, team, forceHooks)
 			hook.Run("TTTPlayerSetColor", self)
 		end
 	end
+
+	if SERVER then 
+		local role_new = roles.GetByIndex(newSubrole)
+		local role_old = roles.GetByIndex(oldSubrole)
+		role_old:RemoveRoleLoadout(self, true)
```
Comment (Alf21) [inline gamemodes/terrortown/gamemode/shared/sh_player_ext.lua]: Missing empty line in front

**Comment by Alf21** [comment]
It's a nice solution anyways

**Pull request merged**